### PR TITLE
Using the MutateRowsRequest api for buffered mutator.

### PR DESCRIPTION
**Pull request by sduskis**

### CODE REVIEW

File: bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataClient.java [hunk 1]
```diff
@@ -48,6 +50,18 @@
   ListenableFuture<Empty> mutateRowAsync(MutateRowRequest request);
 
   /**
+   * Mutate a row atomically.
+   */
+  MutateRowsResponse mutateRows(MutateRowsRequest request) throws ServiceException;
+
+  /**
+   * Mutate a row atomically, returning a Future that will finish when
```
Comment (garye) [inline bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataClient.java]: Copy and paste? "Mutate a row atomically"

File: bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java [hunk 2]
```diff
@@ -174,6 +205,13 @@ public CheckAndMutateRowResponse checkAndMutateRow(CheckAndMutateRowRequest requ
       IS_RETRYABLE_MUTATION, request);
   }
 
+  private ClientCall<MutateRowsRequest, MutateRowsResponse>
+      createMutateRowsCall(MutateRowsRequest request) {
+    expandPoolIfNecessary(this.bigtableOptions.getChannelCount());
```
Comment (garye) [inline bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java]: This gets called by blocking and async versions of mutateRows, can we move expandPoolIfNecessary to the mutateRowsAsync()?

File: bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/AsyncExecutor.java [hunk 3]
```diff
@@ -46,6 +48,8 @@
   public static final long ASYNC_MUTATOR_MAX_MEMORY_DEFAULT =
       (long) Math.min(1 << 30, (Runtime.getRuntime().maxMemory() * 0.1d));
 
+  static final long MAX_SIZE = 1 << 21;
+  static final int MAX_COUNT = 50000;
```
Comment (garye) [inline bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/AsyncExecutor.java]: Are MAX_SIZE and MAX_COUNT used? If so a less generic name would be helpful.

File: bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/AsyncExecutor.java [hunk 4]
```diff
@@ -262,6 +306,21 @@ public void flush() throws IOException {
     LOG.trace("Done flushing");
   }
 
+  public ListenableFuture<MutateRowsResponse> mutateRowsAsync(BulkMutation bulkMutation)
```
Comment (garye) [inline bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/AsyncExecutor.java]: Should have a method comment like (most of) the rest of the public methods in this class.

File: bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/BulkMutation.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+import com.google.protobuf.Empty;
+import com.google.rpc.Status;
+
+import io.grpc.StatusException;
+
+public class BulkMutation {
```
Comment (garye) [inline bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/BulkMutation.java]: Nit: Class comment would be helpful

File: bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/BulkMutation.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    return builder.build();
+  }
+
+  public void addBulkFutureListener(ListenableFuture<MutateRowsResponse> bulkFuture) {
+    FutureCallback<MutateRowsResponse> callback = new FutureCallback<MutateRowsResponse>() {
+      @Override
+      public void onSuccess(MutateRowsResponse result) {
+        Iterator<Status> statuses = result.getStatusesList().iterator();
+        Iterator<SettableFuture<Empty>> entries = futures.iterator();
+        while(entries.hasNext() && statuses.hasNext()) {
```
Comment (garye) [inline bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/BulkMutation.java]: space after while

File: bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/BulkMutation.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+          setStatus(entries.next(), statuses.next());
+        }
+        // TODO: better handling of these cases?
+        while (entries.hasNext()) {
+          entries.next().setException(io.grpc.Status.UNKNOWN
+              .withDescription("Mutation does not have a status").asException());
+        }
+        if (statuses.hasNext()) {
+          int count = 0;
+          while(statuses.hasNext()) {
```
Comment (garye) [inline bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/BulkMutation.java]: space after while

**Comment by sduskis** [comment]
I previously had a merge issue.  I found the good version of the classes.  Please review.

File: bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/BulkMutation.java [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+  static final long MAX_SIZE = 1 << 21;
+  static final int MAX_COUNT = 50000;
+
+  private final List<SettableFuture<Empty>> futures = new ArrayList<>();
+  private final MutateRowsRequest.Builder builder;
+  private long size;
+
+  public BulkMutation(String tableName) {
+    this.builder = MutateRowsRequest.newBuilder().setTableName(tableName);
+    size = MutateRowsRequest.newBuilder().setTableName(tableName).build().getSerializedSize();
```
Comment (garye) [inline bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/BulkMutation.java]: this.size for consistency

File: bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/BulkMutation.java [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+  }
+
+  public synchronized SettableFuture<Empty> add(MutateRowRequest request) {
+    SettableFuture<Empty> future = SettableFuture.create();
+    futures.add(future);
+    MutateRowsRequest.Entry entry = MutateRowsRequest.Entry.newBuilder()
+      .setRowKey(request.getRowKey())
+      .addAllMutations(request.getMutationsList())
+      .build();
```
Comment (garye) [inline bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/BulkMutation.java]: Not sure how expensive the build() is but might want to do that work outside of a synchronized block.
Also the class comment indicates that the class is not thread safe but this method is synchronized.

File: bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/BulkMutation.java [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+      @Override
+      public void onFailure(Throwable t) {
+        handleException(t);
+      }
+    };
+    Futures.addCallback(bulkFuture, callback);
+  }
+
+  public void handleException(Throwable t) {
+    for(SettableFuture<Empty> future : futures) {
```
Comment (garye) [inline bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/BulkMutation.java]: Space after for

File: bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/async/TestAsyncExecutor.java [hunk 11]
```diff
@@ -21,6 +21,7 @@
 import java.util.ArrayList;
 import java.util.List;
 import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
```
Comment (garye) [inline bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/async/TestAsyncExecutor.java]: Looks like none of these new imports are actually used and could be cleaned up

File: bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/BulkMutation.java (hunk 9 above)
Comment (sduskis) [inline bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/BulkMutation.java]: I removed the synchronized indicator.  Calling classes will have to make sure this is thread safe

Comment (sduskis) [inline bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/BulkMutation.java]: BigtableBufferedMutator already does synchronization around this.

File: bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java [hunk 12]
```diff
@@ -157,6 +177,17 @@ public Empty mutateRow(MutateRowRequest request) throws ServiceException {
   }
 
   @Override
+  public MutateRowsResponse mutateRows(MutateRowsRequest request) throws ServiceException {
+    return clientCallService.blockingUnaryCall(createMutateRowsCall(request), request);
+  }
+
+  @Override
+  public ListenableFuture<MutateRowsResponse> mutateRowsAsync(MutateRowsRequest request) {
+    expandPoolIfNecessary(this.bigtableOptions.getChannelCount());
```
Comment (garye) [inline bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java]: In the implementation of expandPoolIfNecessary, should something more serious than LOG.info be used if the pool can't be resized?

File: bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/async/TestBulkMutation.java [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+import com.google.protobuf.ByteString;
+import com.google.protobuf.Empty;
+import com.google.rpc.Status;
+import com.google.rpc.Status.Builder;
+
+/**
+ * Tests for {@link BulkMutation}
+ */
+@RunWith(JUnit4.class)
+public class TestBulkMutation {
```
Comment (garye) [inline bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/async/TestBulkMutation.java]: Like you mentioned, might be good to have a test that covers the unexpected entry count case.

File: bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableBufferedMutator.java [hunk 14]
```diff
@@ -134,15 +141,16 @@ public void run() {
...
-            if (operation == null || operation.isCloseMarker) {
+            if (operation == null) {
+              break;
+            } else if (operation == SHUTDOWN_MARKER) {
               break;
             }
             operation.run();
           } catch (InterruptedException e) {
-            Thread.interrupted();
             LOG.info("Interrupted. Shutting down mutationRunnable.");
```
Comment (garye) [inline bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableBufferedMutator.java]: mutationRunnable seems like an outdated name

File: bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableBufferedMutator.java [hunk 15]
```diff
@@ -134,15 +141,16 @@ public void run() {
       try {
         while (!executorService.isShutdown()) {
           try {
-            MutationOperation operation =
-                mutationsToBeSent.poll(MUTATION_TO_BE_SENT_WAIT_MS, TimeUnit.MILLISECONDS);
+            Runnable operation =
+                asyncOperationsQueue.poll(MUTATION_TO_BE_SENT_WAIT_MS, TimeUnit.MILLISECONDS);
             // The operation can be null if a timeout occurs.
-            if (operation == null || operation.isCloseMarker) {
+            if (operation == null) {
```
Comment (garye) [inline bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableBufferedMutator.java]: More readable with an or? if (operation == null || operation == SHUTDOWN_MARKER) {

File: bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableBufferedMutator.java [hunk 16]
```diff
@@ -283,21 +313,56 @@ public void mutate(final Mutation mutation) throws IOException {
...
     try {
       initializeAsyncMutators();
-      long operationId = heapSizeManager.registerOperationWithHeapSize(mutation.heapSize());
-      MutationOperation operation = new MutationOperation(mutation, operationId, false);
-      if (executorService != null && options.getAsyncMutatorCount() > 0) {
-        mutationsToBeSent.add(operation);
+      Runnable operation = null;
+      if (options.useBulkApi() && (mutation instanceof Put || mutation instanceof Delete)) {
+        // TODO: Do this logic asynchronously. 
+        synchronized (bulkMutationLock) {
```
Comment (garye) [inline bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableBufferedMutator.java]: I think this new behavior makes the comment on mutate() not exactly true - it only blocks if those two cases are true and it's not doing a bulk mutation.

File: bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/async/TestBulkMutation.java (hunk 13 above)
Comment (sduskis) [inline bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/async/TestBulkMutation.java]: done

Comment (garye) [inline bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/async/TestBulkMutation.java]: nice!

File: bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableBufferedMutator.java (hunk 16 above)
Comment (sduskis) [inline bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableBufferedMutator.java]: The comment is still true with bulk mutations.  AsyncExecutor.mutateRowsAsync() will make sure that happens.

File: bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/TestBigtableBufferedMutator.java [hunk 17]
```diff
@@ -157,21 +168,51 @@ public void testInvalidPut() throws Exception {
...
       underTest.mutate(SIMPLE_PUT);
-      verify(client, times(1)).mutateRowAsync(any(MutateRowRequest.class));
+      verify(mockClient, times(1)).mutateRowAsync(any(MutateRowRequest.class));
       Assert.assertTrue(underTest.hasInflightRequests());
       completeCall();
       Assert.assertFalse(underTest.hasInflightRequests());
     }
   }
 
+  @Test
```
Comment (garye) [inline bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/TestBigtableBufferedMutator.java]: Might be beneficial to have a test, or enhance this test, to cover the case where multiple mutates / multiple mutateRows calls get sent to test the resetting of BulkMutation state.

File: bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableBufferedMutator.java (hunk 16 above)
Comment (garye) [inline bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableBufferedMutator.java]: Ah, I see. Can bulkMutation.add() block, i.e. in the not full yet case? 

Also the comment for mutate() has "Being a Mutation", probably should be "Begin a mutation"?

File: bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java (hunk 12 above)
Comment (sduskis) [inline bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java]: The pool expansion is a  performance enhancement and not a major functionality, so it's questionable.  We can discuss the severity offline.

File: bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/TestBigtableBufferedMutator.java (hunk 17 above)
Comment (sduskis) [inline bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/TestBigtableBufferedMutator.java]: done.

**Comment by garye** [comment]
LGTM

**Pull request merged**